common/main/java/com/couchbase/lite/AbstractReplicator.java

```diff
@@ -382,15 +386,43 @@ public void removeChangeListener(@NonNull ListenerToken token) {
         }
     }
 
-    // This is a workaround: closing at the wrong time (while a push filter is running?) might even
-    // cause a crash.  The issue it addresses is that when a replicator is closed it is detached from
-    // its from LiteCore and receives no further status updates. If it is not already in the STOPPED
-    // state, it will never be in the stopped state.  That means that a database that holds a reference
-    // to it can never close.
-    // Let's just tell the db forget about it.
+    // I've thought a lot about how to implement this.  The problem is that you cannot, fundamentally,
+    // close a replicator(discard its resources before this menthod returns). If it is not stopped,
+    // it will have to be stopped.  I came up with several possibilities.
+    // 1) close it if it is stopped; stop it if it is running.
+    // 2) throw an exception on the attempt to close a running replicator.
+    // 3) tell it to stop and forget it, without verifying that it actually goes away.
+    // 4) Make replicators not Autoclosable.
+    // #4 is very tempting.  Replicators will never be used in a try-with-resources block.
+    // Replicators *do*, however, dominate a LiteCore object that must be freed explicitly.
+    // Making them, transitively, Autoclosable seems correct.
+    // #2 seems to put an excessive burden on client code: it is a bomb that explodes unless you
+    // wait around and pay very close attention to the state of every replicator.
+    // I dislike #1 because it puts the replicator into an odd state in which it is still kinda
+    // half alive.  The client closed it but it isn't quite gone yet.
+    // I have chosen #3.  My rational is that, if you create and start a replicator and then null out
+    // all references to it, this is exactly what will happen: the replicator will be forgotten and
+    // eventually, GCed, told to stop, and freed.  This method does precisely that: best effort to
+    // release the resources and then forget about them.
```

Comment:
My impression that close() method has synchronize like .NET's dispose() or stream.close(). So when calling close(), the replicator will be stopped and its resource will be released. So maybe #1 with synchronous behavior to ensure that the replicator is stopped?

 Also, I think the Java and .NET could be aligned on the behavior and implementation. CC: @borrrden.

Reply:
I see.  Sure, I could do that: no zombie state.  It might take minutes, though.  I have three arguments against that idea:
1) I think Java devs would find it quite surprising if a `close()` method took minutes to run.  Not a blocker, but surprising.
2) There is no way that we could wait for the close, in the finalizer.  That would be catastrophic.  If close is going to do what GC does (which I claim it should), it should not wait.
3) the reason you would wait is in case that the replicator did not close cleanly.  But what would client code do with that information?  I submit that the best effort attempt will usually work, and that when it doesn't work, there's nothing that the user can do, anyway.

The only argument I can see for waiting is that client code wants to do something that they can only do, once the replicator is definitely stopped.  There is a perfectly straightforward way to do that, right now without the `close` method: attach a listener and wait for the stopped state.